### PR TITLE
Convert Alembic importer to reader

### PR DESCRIPTION
**Pull request by Meakk**

### CODE REVIEW

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    if (pmesh.getSchema().getNumSamples() > 0)
+    {
+      pmesh.getSchema().get(samp);
+
+      Alembic::AbcGeom::P3fArraySamplePtr positions = samp.getPositions();
+      Alembic::AbcGeom::Int32ArraySamplePtr indices = samp.getFaceIndices();
+      Alembic::AbcGeom::Int32ArraySamplePtr counts = samp.getFaceCounts();
+
+      size_t P_size = positions->size();
+      size_t counts_size = counts->size();
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:41]: perfect time to follow guidelines

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+      for (size_t i = 0; i < P_size; i++)
+      {
+        points->InsertNextPoint(
+          positions->get()[i].x, positions->get()[i].y, positions->get()[i].z);
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:46]: switching to SetPoint should be trivial

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      }
+
+      size_t face_index = 0;
+      for (size_t i = 0; i < counts_size; i++)
+      {
+        auto polyface_vertex_count = counts->get()[i];
+        polys->InsertNextCell(polyface_vertex_count);
+        for (auto j = 0; j < polyface_vertex_count; j++)
+        {
+          polys->InsertCellPoint(indices->get()[face_index++]);
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:56]: using a faster API should be trivial

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+      append->AddInputData(ProcessIPolyMesh(polymesh));
+      nextParentObject = polymesh;
+    }
+
+    // Recursion
+    if (nextParentObject.valid())
+    {
+      for (size_t i = 0; i < nextParentObject.getNumChildren(); ++i)
+      {
+        IterateIObject(append, nextParentObject, nextParentObject.getChildHeader(i));
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:89]: this

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    }
+  }
+
+  void ImportRoot(vtkAppendPolyData* append)
+  {
+    Alembic::Abc::IObject top = this->Archive.getTop();
+
+    for (size_t i = 0; i < top.getNumChildren(); ++i)
+    {
+      IterateIObject(append, top, top.getChildHeader(i));
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:100]: this

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    }
+  }
+
+  void ReadScene(const std::string& filePath)
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:104]: rename to PrepareArchive ?

File: plugins/alembic/module/vtkF3DAlembicReader.cxx [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+//----------------------------------------------------------------------------
+int vtkF3DAlembicReader::RequestData(
+  vtkInformation*, vtkInformationVector**, vtkInformationVector* outputVector)
+{
+  vtkPolyData* output = vtkPolyData::GetData(outputVector);
+
+  this->Internals->ReadScene(this->FileName);
+
+  vtkNew<vtkAppendPolyData> append;
```
Comment (mwestphal) [inline plugins/alembic/module/vtkF3DAlembicReader.cxx:134]: does that work when working with textures ?

**Comment by mwestphal** [comment]
doc update missing

File: plugins/alembic/module/vtkF3DAlembicReader.cxx (hunk 7 above)
Comment (Meakk) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:134]: yes it will use the same textures for all submeshes.

Comment (mwestphal) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:139]: is that ok ? would be beneficial to have @nyue input on this

Comment (Meakk) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:139]: Well, that's the case with all geometry-only formats.  
As far as I know, Alembic supports texture coordinates but no materials. The texture coordinates are supposed to be used in a later stage in Blender/Maya or any final rendering software by crafting materials there.  
The goal of f3d here is to provide a way to visualize these texture coordinates by using a checkboard texture for example.

Comment (nyue) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:139]: Technically, Alembic does have a [prototype](https://github.com/alembic/alembic/tree/master/lib/Alembic/AbcMaterial) material support but most studio working with Alembic have their own way to deal with what they call *Look Development* in their pipeline and the industry is gravitating towards [MaterialX](https://materialx.org/).

I believe for a start, what @Meakk solution provides is sufficient.

Comment (mwestphal) [reply plugins/alembic/module/vtkF3DAlembicReader.cxx:139]: Alright then!

**Comment by Meakk** [comment]
> doc update missing

what doc?

**Comment by mwestphal** [comment]
> > doc update missing
> 
> what doc?

I was thinking about this:

https://github.com/f3d-app/f3d/blob/master/doc/user/USAGE.md?plain=1#L28

But it was never listed as a full scene format anyway. soo all good.

**Review by mwestphal — APPROVED** [review]

**Pull request merged**